internal/api/server.go

```diff
@@ -255,3 +259,52 @@ func (s *server) Serve(ctx context.Context, l net.Listener) error {
 		return err
 	}
 }
+
+func (s *server) newDashboardRequestHandler() (http.HandlerFunc, error) {
+	uiFS := fs.FS(ui)
+	uiFS, err := fs.Sub(uiFS, "ui")
+	if err != nil {
+		return nil, fmt.Errorf("error initializing UI file system: %w", err)
+	}
+
+	handler := http.FileServer(http.FS(uiFS))
+	return func(w http.ResponseWriter, req *http.Request) {
+		if req.Method != "GET" {
+			http.Error(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
+			return
+		}
+
+		path := filepath.Clean(req.URL.Path)
+		if path == "/" {
+			path = "index.html"
+		}
+		path = strings.TrimPrefix(path, "/")
+
+		f, err := uiFS.Open(path)
+		if err != nil {
+			// When the path doesn't match an embedded file, serve index.html
+			if os.IsNotExist(err) {
+				httputil.SetNoCacheHeaders(w)
+				http.ServeFileFS(w, req, uiFS, "index.html")
+			}
+			http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+			return
+		}
+
+		info, err := f.Stat()
+		if err != nil {
+			http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+			return
+		}
+
+		// When the path matches a directory, serve index.html (prevents
+		// enumerating files in the directory)
+		if info.IsDir() {
+			httputil.SetNoCacheHeaders(w)
+			http.ServeFileFS(w, req, uiFS, "index.html")
+			return
+		}
+
+		handler.ServeHTTP(w, req)
+	}, nil
+}
```

Comment:
```suggestion
func newDashboardRequestHandler() (http.HandlerFunc, error) {
	const indexHTML = "index.html"

	uiFS := fs.FS(ui)
	uiFS, err := fs.Sub(uiFS, "ui")
	if err != nil {
		return nil, fmt.Errorf("error initializing UI file system: %w", err)
	}

	handler := http.FileServer(http.FS(uiFS))
	return func(w http.ResponseWriter, req *http.Request) {
		if req.Method != http.MethodGet {
			http.Error(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
			return
		}

		path := filepath.Clean(req.URL.Path)
		if path == "/" {
			http.ServeFileFS(w, req, uiFS, indexHTML)
			return
		}

		f, err := uiFS.Open(strings.TrimPrefix(path, "/"))
		if f != nil {
			defer f.Close()
		}
		if os.IsNotExist(err) {
			// When the path doesn't match an embedded file, serve index.html
			httputil.SetNoCacheHeaders(w)
			http.ServeFileFS(w, req, uiFS, indexHTML)
			return
		}
		if err != nil {
			http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
			return
		}

		// If we get to here, the path exists in the embedded file system

		info, err := f.Stat()
		if err != nil {
			http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
			return
		}
		if info.IsDir() {
			// Serve index.html to prevent enumerating files in the directory
			httputil.SetNoCacheHeaders(w)
			http.ServeFileFS(w, req, uiFS, indexHTML)
			return
		}

		// Path is a file
		handler.ServeHTTP(w, req)
	}, nil
}
```

Reply:
Ah, forgot to `Close()` the file 🤦 

In this section of the suggestion, is there a particular reason to not set the cache headers?

```go
if path == "/" {
	http.ServeFileFS(w, req, uiFS, indexHTML)
	return
}
```

Reply:
@krancour I took the suggestion, with the slight change of adding the `httputil.NoCacheHeaders(w)` to that one line.

Reply:
🤦‍♂️ Thank you!